### PR TITLE
IDS - xmlschema, restrictions and requirements

### PR DESCRIPTION
**Pull request by atomczak**
- Switched from xml.dom to xmlschema.
- improved requirements with proper validation
- implemented missing restrictions

### CODE REVIEW

File: src/ifcopenshell-python/ifcopenshell/ids.py [hunk 1]
```diff
@@ -97,17 +105,23 @@ class classification(facet):
     message = "a classification reference '%(value)s' from '%(system)s'"
 
     def __call__(self, inst, logger):
+        #TODO Location: 'type'/'instance'/'any'
         refs = []
         for association in inst.HasAssociations:
             if association.is_a("IfcRelAssociatesClassification"):
                 cref = association.RelatingClassification
-                refs.append((cref.ReferencedSource.Name, cref.ItemReference))
+                refs.append((cref.ReferencedSource.Name, cref.Identification))  # before was .ItemReference instead of .Identification
```
Comment (aothms) [inline src/ifcopenshell-python/ifcopenshell/ids.py:113]: This still needs to be handled in a schema-agnostic way right? ItemReference for 2x3, Identification for IFC4.

**Comment by aothms** [comment]
Looks really good I think. Couple of small remarks, some only for future work.

File: src/ifcopenshell-python/ifcopenshell/ids.py (hunk 1 above)
Comment (atomczak) [reply src/ifcopenshell-python/ifcopenshell/ids.py]: Of course. Left comment to myself and forgot to reflect both schemas in code. Just pushed a commit with that change.

**Comment by aothms** [comment]
Thanks! Really cool to see the progress :)

**Comment by aothms** [comment]
PS we're running the tests on Travis CI now as well so we get notified in case something breaks

afbeelding

https://app.travis-ci.com/github/IfcOpenShell/IfcOpenShell/builds/232381022#L1441

**Pull request merged**